### PR TITLE
Add multiplayer

### PR DESCRIPTION
**Pull request by lalalune**
This fixes a myriad of multiplayer bugs and doesn't seem to harm single player with any of the items or abilities.

Avatar movement works in multplayer. Vehicles work in multiplayer. Wearing still needs some work -- 'wearupdate' is not being called in multiplayer, and so it is not updating. A few small bugs here and there which need to next be handled (issues on our Github).

### CODE REVIEW

File: character-controller.js [hunk 1]
```diff
@@ -541,10 +542,10 @@ class StatePlayer extends PlayerBase {
   }
   // serializers
   getPosition() {
-    return this.playerMap.get('position') ?? [0, 0, 0];
+    return this.position.toArray() ?? [0, 0, 0];
```
Comment (avaer) [inline character-controller.js:545]: Is this running every frame? If so this is not acceptable garbage generation.

File: world.js [hunk 2]
```diff
@@ -92,9 +90,6 @@ world.connectRoom = async u => {
       wsrtc.removeEventListener('init', init);
       
       localPlayer.bindState(state.getArray(playersMapName));
-      if (mediaStream) {
```
Comment (avaer) [inline world.js:95]: Does this lose mic functionality?

File: character-physics.js [hunk 3]
```diff
@@ -157,16 +157,29 @@ class CharacterPhysics {
...
+        // NOTE: This works, but is probably not very performant for more than a few riders
+        // TODO: Optimize this
+        let rideMesh = null;
+        controlledApp.glb.scene.traverse(o => {
+          if (rideMesh === null && o.isSkinnedMesh) {
+            rideMesh = o;
+          }
+        });
+
+        // NOTE: We had a problem with sending the entire bone in the message buffer, so we're just sending the bone name
```
Comment (avaer) [inline character-physics.js:179]: What does "sending the bone" mean? As far as I know we never send bones anywhere.

**Comment by alisaad673** [comment]
### Single player Issues:
### **1-Drag & dropping assets gives error on both "drop to world" and "equip to self". In both cases it gives an error and asset appears in app. We can wear the asset though.**
### **Drop to world:**
single play drop to world err

https://user-images.githubusercontent.com/35926530/162798617-1f623a64-7b88-4635-9c7a-137241acf53c.mp4


### **Equip to self:**
single play equip to self err

https://user-images.githubusercontent.com/35926530/162798599-0f78df19-29c0-45ab-811c-735c0936b57d.mp4


###  **2-Trying to change avatar in Character select panel gives an error and we are unable to wear that avatar**
character select change avatar dosnt works

https://user-images.githubusercontent.com/35926530/162799199-f3b365ff-2966-4d65-a970-07e1a0a467be.mp4


### **3-App Crashes while we try to equip the asset from tokens in character panel. It gives an error first while we drag & drop it. Then if we try to wear it, it crashes the app.**

equip asset crash from charactr panel

https://user-images.githubusercontent.com/35926530/162799779-4fcafb3a-ea71-4029-a4dd-ee307a872ce9.mp4


### Multiplayer Issues:

All the above mentioned issues are there too in multiplayer, Additionally these issues are there too in multiplayer.

### **5-Some of the times one avatar seems to be sliding instead of walking or running in one or both of the instances. Refreshing the app fixes this.**

https://user-images.githubusercontent.com/35926530/162800280-f3a30af8-eea1-4684-82cd-27dfb5c97a09.mp4

### **6-On the initial load, Avatars passes through other avatar, Refreshing both the instances fixes this issue.**

https://user-images.githubusercontent.com/35926530/162800586-0df164ef-656c-4413-85ed-be0ae468867c.mp4

### **7-euipping fox shows weird behavior in other instance.**

https://user-images.githubusercontent.com/35926530/162801321-d3bdb635-aa5f-4124-9138-9df3712dc70f.mp4

### **8-While we equip fox in one instance, Though it shows weird behavior in other instance but we are still able to equip the same (already equipped by 1st instance) fox here too, So it becomes a weird response with fox as shown in video below.**

https://user-images.githubusercontent.com/35926530/162802697-46ac3218-8457-4047-b231-9a9eef553d44.mp4

### **9-Equipping Sword in one instance doesn't appears in other instance. Has to be refreshed. Even after refreshing, the sword combat doesn't replicates in other instance.**

https://user-images.githubusercontent.com/35926530/162801749-a570edba-c8b0-49eb-8ea3-a8c0aba02c97.mp4

### 10-Refreshing while we have equipped vehicle in one instance makes the behavior weird after refreshing as shown in video(The Avatar only seems to be replicating the animation only without the vehicle model)

https://user-images.githubusercontent.com/35926530/162802142-2c09348d-65bc-4d82-88a1-257af91db20d.mp4

### 11-Trying to equip dragon, gives this error and crashes the app.
equip dragon crash err

https://user-images.githubusercontent.com/35926530/162802293-6fe39eab-fd91-4a5e-9dc6-bc1618206e9e.mp4

File: world.js (hunk 2 above)
Comment (lalalune) [reply world.js:95]: No, and it was preventing multiplayer from working at all there, the code is in character-controller.js now
image

**Comment by alisaad673** [comment]
**Reloading bugs:**

**1-On the first load the avatars appear to be sliding instead of walking in the other instance. Refreshing both the instances fixes the bug.**

https://user-images.githubusercontent.com/35926530/165792104-e8a3b5b5-a71d-4b29-b960-008e7d7cebdd.mp4


**2-If we don't refresh the instances, the multiplayer doesn't works fine and gives various bugs such as sword combat doesn't interpolates, avatar sliding etc.
3-If we refresh, There are certain bugs too such as the standby vrm disappears in both instances once we refresh.**


**Dash:**

**1-While we try to do dash(naruto run(Shift+double W)) we dont get the dash effect(glow) in the other instance.**

https://user-images.githubusercontent.com/35926530/165794065-16f80477-346d-43fb-afe4-4d4fa8d6bd40.mp4



**Fox:**

**1-Equipping the pet in one instance works fine as the pet is walking in the instance it is equipped in, But in the other instance, pet appears to be sliding instead of walking.**

https://user-images.githubusercontent.com/35926530/165792370-dc89f872-2142-4048-b9c2-ea2aae45b8bf.mp4



**2-Also when we equip the pet in one instance, the pet's face points to the avatar which has equipped it. But in the other instance the pet's face points to the standing native avatar instead of the one that equipped it.**

https://user-images.githubusercontent.com/35926530/165792438-8b3291c1-5459-4fb4-84e5-248cd1e601eb.mp4



**3-If we equip fox first and then we try to equip the sword, The fox also goes to the position of sword on spine and looks too weird.**


https://user-images.githubusercontent.com/35926530/165792488-5c55a1e5-da11-4166-808c-5b1d22a7dbc7.mp4



**Sword:**

**1-Combat works fine. But We are unable to damage the other player. We can damage the objects in there.**


**Vehicles:**
**1-Equipping the vehicle sometimes work fine but in some cases, it gives error in one of the instances and it crashes the app.**
errwhile equiping vehicle


**Dragon:**
**1-Equipping Dragon gives this error and crashes the instance**
equip dragon

https://user-images.githubusercontent.com/35926530/165792746-0a17524f-867e-4d19-acc6-63c9d1f5d724.mp4



**Switching scene:**
**1-Trying to switch scene gets us out of multiplayer.**

**Comment by lalalune** [comment]
This can be reviewed at:
https://webaverse.n3xus.city:3000/?src=.%2Fscenes%2Fstreet.scn&room=qa
https://user-images.githubusercontent.com/18633264/170807820-93ad5c6d-cde2-4213-ac57-a5aa9e0f04bd.mp4

**Comment by alisaad673** [comment]
As far as the basic features are concerned:
1-Interpolation has a jitter in it sometimes(I am not sure on why is it as sometime it seems fine.) Even though the jitter is not there, there can be seen a slight delay in interpolation.

https://user-images.githubusercontent.com/35926530/171045458-873bc73d-c7c2-4ec8-8ab3-f800ec88d1f7.mp4


2-There is no Sound for chat.(The Character has the mouth movements but doesn't speaks with sound)

https://user-images.githubusercontent.com/35926530/171045909-9b85aecd-e5f6-4b87-8d7a-994edff70058.mp4

3-When we hover over characters in character select panel, It throws this error

image
https://user-images.githubusercontent.com/35926530/171047205-a95ab7a2-5178-4078-a933-b2842f695c00.mp4


Working:
-walk/run/fly/crouch/dance/dash interpolation works
-Sword/bow combat works
-Chat works but no sound
-Creating room works
-Drag and drop of vrms and glbs work

**Comment by lalalune** [comment]
@alisaad673 thanks for the feedback. Fixing #3050 
I think that your first comment is because we are capping the interpolation window at 100ms, so any time it jumps outside of that it causes a stutter. I would expect with interpolation that we have a very slight delay between sides (1/10th second basically)
will look into it and have fixes shortly

**Pull request closed without merging**